processing/src/main/java/io/druid/query/lookup/LookupReferencesManager.java

```
@@ -52,333 +55,400 @@
+  //for unit testing only
+  private final boolean testMode;
 
   @Inject
-  public LookupReferencesManager(LookupConfig lookupConfig, final @Json ObjectMapper objectMapper)
+  public LookupReferencesManager(LookupConfig lookupConfig, @Json ObjectMapper objectMapper)
+  {
+    this(lookupConfig, objectMapper, false);
+  }
+
+  @VisibleForTesting
+  LookupReferencesManager(LookupConfig lookupConfig, ObjectMapper objectMapper, boolean testMode)
   {
     if (Strings.isNullOrEmpty(lookupConfig.getSnapshotWorkingDir())) {
       this.lookupSnapshotTaker = null;
     } else {
       this.lookupSnapshotTaker = new LookupSnapshotTaker(objectMapper, lookupConfig.getSnapshotWorkingDir());
     }
+
+    this.testMode = testMode;
   }
 
   @LifecycleStart
   public void start()
   {
-    startStopLock.writeLock().lock();
+    if (!lifecycleLock.canStart()) {
+      throw new ISE("can't start.");
+    }
+
     try {
-      if (!started.getAndSet(true)) {
-        if (lookupSnapshotTaker != null) {
-          final List<LookupBean> lookupBeanList = lookupSnapshotTaker.pullExistingSnapshot();
-          for (LookupBean lookupBean : lookupBeanList) {
-            this.put(lookupBean.name, lookupBean.factory);
-          }
-        }
-        LOGGER.info("Started lookup factory references manager");
+      LOG.info("LookupReferencesManager is starting.");
+
+      loadSnapshotAndInitStateRef();
+
+      if (!testMode) {
+        mainThread = Execs.makeThread(
+            "LookupReferencesManager-MainThread",
+            new Runnable()
+            {
+              @Override
+              public void run()
+              {
+                try {
+
+                  if (!lifecycleLock.awaitStarted()) {
+                    LOG.error("WTF! lifecycle not started, lookup update notices will not be handled.");
+                    return;
+                  }
+
+                  while (!Thread.interrupted() && lifecycleLock.awaitStarted(1, TimeUnit.MILLISECONDS)) {
```

Note that `lifecycleLock.awaitStarted` swallows interrupted exceptions into runtime exceptions and clears interrupted state of thread.